reference/conanfile/tools/cmake.rst

```diff
@@ -192,11 +184,108 @@ For single-configuration build systems:
     $ cmake --build .  # or just "make"
 
 
-Conan is able to generate a toolchain file for different systems. In the
-following sections you can find more information about them:
+Extending and customizing CMakeToolchain
+++++++++++++++++++++++++++++++++++++++++
+
+Since Conan 1.36, ``CMakeToolchain`` implements a powerful capability for extending and customizing the resulting toolchain file.
+
+The following predefined blocks are available:
+
+- ``generic_system``: Defines ``CMAKE_GENERATOR_PLATFORM``, ``CMAKE_GENERATOR_TOOLSET``, ``CMAKE_C_COMPILER``,``CMAKE_CXX_COMPILER`` and ``CMAKE_BUILD_TYPE``
+- ``android_system``: Defines ``ANDROID_PLATFORM``, ``ANDROID_STL``, ``ANDROID_ABI`` and includes ``CMAKE_ANDROID_NDK/build/cmake/android.toolchain.cmake``
+  where CMAKE_ANDROID_NDK comes defined in ``tools.android:ndk_path``
+- ``ios_system``: Defines ``CMAKE_SYSTEM_NAME``, ``CMAKE_SYSTEM_VERSION``, ``CMAKE_OSX_ARCHITECTURES``, ``CMAKE_OSX_SYSROOT`` for Apple systems.
+- ``find_paths``: Defines ``CMAKE_FIND_PACKAGE_PREFER_CONFIG``, ``CMAKE_MODULE_PATH``, ``CMAKE_PREFIX_PATH`` so the generated files from ``CMakeDeps`` are found.
+- ``fpic``: Defines the ``CMAKE_POSITION_INDEPENDENT_CODE`` when there is a ``options.fPIC``
+- ``rpath``: Defines ``CMAKE_SKIP_RPATH`` for OSX
+- ``arch_flags``: Defines C/C++ flags like ``-m32, -m64`` when necessary.
+- ``libcxx``: Defines ``-stdlib=libc++`` flag when necessary as well as ``_GLIBCXX_USE_CXX11_ABI``.
+- ``vs_runtime``: Defines the ``CMAKE_MSVC_RUNTIME_LIBRARY`` variable, as a generator expression for
+  for multiple configurations.
+- ``cppstd``: defines ``CMAKE_CXX_STANDARD``, ``CMAKE_CXX_EXTENSIONS``
+- ``shared``: defines ``BUILD_SHARED_LIBS``
+- ``parallel``: defines ``/MP`` parallel build flag for Visual.
+
+
+Blocks can be customized in different ways:
+
+.. code:: python
+
+    # remove an existing block
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.pre_blocks.remove("generic_system")
+
+    # modify the template of an existing block
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tmp = tc.pre_blocks["generic_system"].template
+        new_tmp = temp.replace(...)  # replace, fully replace, append...
+        tc.pre_blocks["generic_system"].template = new_tmp
+
+    # modify the context (variables) of an existing block
+    def generate(self):
+        tc = CMakeToolchain(self)
+        generic_block = toolchain.pre_blocks["generic_system"]
+
+        def context(self):
+            assert self  # Your own custom logic here
+            return {"build_type": "SuperRelease"}
+        generic_block.context = types.MethodType(context, generic_block)
```

Comment:
Where is this `types` coming from? Is this python syntax or something provided by conan?

Reply:
Yes, this is native python

Reply:
Maybe adding the import above to make it explicit?